ITISFoundation/osparc-control
# Version 0.0.1

# What do these changes do?

Add a base library to be used for control cases. 
**NOTE:** this package was build using [cookiecutter-hypermodern-python](https://cookiecutter-hypermodern-python.readthedocs.io/en/2021.11.26/quickstart.html) for convenience.

# How does it work?

- The `osparc_control` package provides an easy way to define a supported interface by providing a list of `CommandManifest`.
- A `CommandManifest` is the interface(signature) of the supported command.
- Each request from a remote is validated against `CommandManifest`. A `CommandReceived` confirmation is always send to each command. The requester expects this command to be delivered (this is required to avoid the requesting part being stuck awaiting forever if the remote part is not running). A 1 minute interval is defined to detect this situation.
- It supports 3 types of commands defined by `CommandType`:
```python
class CommandType(str, Enum):
    # the command expects no reply
    WITHOUT_REPLY = "WITHOUT_REPLY"

    # the command will provide a reply
    # the user is require to check for the results
    # of this reply
    WITH_DELAYED_REPLY = "WITH_DELAYED_REPLY"

    # the command will return the result immediately
    # and user code will be blocked until reply arrives
    # used for very fast replies (provide data which already exists)
    WITH_IMMEDIATE_REPLY = "WITH_IMMEDIATE_REPLY"
```

# How to test

You require `poetry` on your system, please make sure it is installed.

In a virtualenv install dev dependencies
```bash
make install-dev
```

Run suite with `python 3.6` and `nox`.

```bash
make nox-36
```
If you do not have python 3.6 on your system feel free to run with a different python interpreter by changing the below command (replace `3.6` with your version of python:

```bash
nox -p 3.6
```

# Examples

Please follow the updated `README.md`

# What changed

- examples are now tested 
- replaced `sphynx` with `docsify`
- documentation will be available at https://itisfoundation.github.io/osparc-control/ once this is merged to master

# Related issues
- https://github.com/ITISFoundation/osparc-issues/issues/555
